# Challenge 1: multiple of 3 and 5

If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.

Find the sum of all the multiples of 3 or 5 below 1000

In [1]:
def get_multiples(base, limit):
    multiples = set()
    
    term = 1
    while True:
        result = base * term
        if result >= limit:
            return multiples
        else:
            multiples.add(result)
        term += 1

In [2]:
%%time
limit = 10
sum(get_multiples(3, limit).union(get_multiples(5, limit)))

CPU times: user 40 µs, sys: 0 ns, total: 40 µs
Wall time: 43.2 µs


23

In [3]:
%%time
limit = 1_000
sum(get_multiples(3, limit).union(get_multiples(5, limit)))

CPU times: user 145 µs, sys: 32 µs, total: 177 µs
Wall time: 181 µs


233168

# Challenge 2: Even Fibonacci numbers

Each new term in the Fibonacci sequence is generated by adding the previous two terms. By starting with 1 and 2, the first 10 terms will be:

1, 2, 3, 5, 8, 13, 21, 34, 55, 89, ...

By considering the terms in the Fibonacci sequence whose values do not exceed four million, find the sum of the even-valued terms.


In [4]:
generated_number_count = 11

In [5]:
def fibonacci_generator():
    a = 0
    b = 1
    while True:
        yield a
        current_a = a
        a = b
        b = b + current_a

In [6]:
number_count = 0
fib_number_list = list()
for fib_number in fibonacci_generator():
    fib_number_list.append(fib_number)
    number_count += 1
    if number_count > generated_number_count:
        break
print(fib_number_list)

[0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89]


In [7]:
def even_fibonacci_generator():
    a = 2
    b = 0
    while True:
        yield a
        current_a = a
        a = 4 * current_a + b
        b =  current_a

In [8]:
number_count = 0
fib_number_list = list()
for fib_number in even_fibonacci_generator():
    fib_number_list.append(fib_number)
    number_count += 1
    if number_count > generated_number_count:
        break
print(fib_number_list)

[2, 8, 34, 144, 610, 2584, 10946, 46368, 196418, 832040, 3524578, 14930352]


In [9]:
%%time 
even_sum = 0
for number in even_fibonacci_generator():
    if number > 4_000_000:
        break
    even_sum += number
print(even_sum)

4613732
CPU times: user 101 µs, sys: 22 µs, total: 123 µs
Wall time: 102 µs


# Challenge 3: Largest prime factor

The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600851475143 ?

In [10]:
import math

def get_prime_factors(n):
    prime_factors = list()
    primes = list()
    search_limit = int(math.sqrt(n)) + 1
    
    remainder = n
    
    for number in range(2, search_limit):
        if remainder % number == 0:
            multiple_of_existing_prime = False
            for prime in primes:
                if number % prime == 0:
                    multiple_of_existing_prime = True
                    break
            if not multiple_of_existing_prime:
                prime_factors.append(number)
                primes.append(number)
                remainder = remainder / number
    return prime_factors

In [11]:
%%time
get_prime_factors(13195)

CPU times: user 28 µs, sys: 0 ns, total: 28 µs
Wall time: 30.5 µs


[5, 7, 13, 29]

In [12]:
%%time
get_prime_factors(600851475143)

CPU times: user 50.5 ms, sys: 277 µs, total: 50.8 ms
Wall time: 49.9 ms


[71, 839, 1471, 6857]

# Challenge 4: Largest palindrome product

A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.

## Notes
First 30 palindromic numbers: 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 11, 22, 33, 44, 55, 66, 77, 88, 99, 101, 111, 121, 131, 141, 151, 161, 171, 181, 191, 202

Max value with 1 digit number: 9x9 = 81 (2 digit number) but there is no palindromic number that can be found with 1 digit number multiplication

Max value with 2 digit number: 99x99 = 9801 (4 digit numbers)

Max value with 3 digit numbers: 999x999 = 998001 (6 digit numbers)

In [13]:
def is_palindromic_number(number):
    number_list = [int(i) for i in str(number)]
    reverse_number_list = number_list[::-1]
    if number_list == reverse_number_list:
        return True
    else:
        return False

In [14]:
%%time
print(f'9009 ? {is_palindromic_number(9009)}')
print(f'1 ? {is_palindromic_number(1)}')
print(f'81 ? {is_palindromic_number(81)}')

9009 ? True
1 ? True
81 ? False
CPU times: user 301 µs, sys: 56 µs, total: 357 µs
Wall time: 240 µs


In [15]:
from functools import reduce
import math

def get_factors(n, digit_count):
    if n == 0:
        return 0
    step = 2 if n % 2 else 1
    
    for factor in set(reduce(list.__add__, ([i, n// i] for i in range(1, int(math.sqrt(n))+1, step) if n % i == 0))):
        if len(str(factor)) == digit_count:
            other_factor = int(n / factor)
            if len(str(other_factor)) == digit_count:
                return (factor, other_factor)
    return None

In [16]:
%%time
get_factors(9009, 2)

CPU times: user 46 µs, sys: 8 µs, total: 54 µs
Wall time: 57.9 µs


(91, 99)

In [17]:
def get_palindromic_number(min_digit_number, max_digit_number, max_digit_count):
    number = max_digit_number ** 2
    while number >= 0:
        if is_palindromic_number(number):
            factors = get_factors(number, max_digit_count)
            if factors:
                return number, factors
        number -= 1
    return None

In [18]:
%%time
get_palindromic_number(0, 9, 1)

CPU times: user 161 µs, sys: 29 µs, total: 190 µs
Wall time: 196 µs


(9, (1, 9))

In [19]:
%%time
get_palindromic_number(10, 99, 2)

CPU times: user 940 µs, sys: 0 ns, total: 940 µs
Wall time: 963 µs


(9009, (91, 99))

In [20]:
%%time
get_palindromic_number(100, 999, 3)

CPU times: user 96.2 ms, sys: 3.94 ms, total: 100 ms
Wall time: 99 ms


(906609, (993, 913))

# Challenge 5: Smallest multiple


2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?


## Notes

Evenly divisible means divisible without remainder

From 1 to 10, we can ignore the following numbers:

* 1
* 2 because divisor of 10, 8, 6, 4
* 3 because divisor of 9, 6
* 4 because divisor of 8
* 5 because divisor of 10

We just need to find the smallest multiple divisible by 10, 9, 8, 7, 6

Biggest multiple is 30240

In [21]:
def get_largest_divisors(min_number, max_number):
    divisors = list()
    for number in range(max_number, min_number - 1, -1):
        found_divisor = True
        for divisor in divisors:
            if divisor % number == 0:
                found_divisor = False
                break
        if found_divisor:
            divisors.append(number)
    return divisors

In [22]:
%%time
get_largest_divisors(1, 10)

CPU times: user 5 µs, sys: 1 µs, total: 6 µs
Wall time: 7.63 µs


[10, 9, 8, 7, 6]

In [23]:
%%time
get_largest_divisors(1, 20)

CPU times: user 0 ns, sys: 8 µs, total: 8 µs
Wall time: 9.54 µs


[20, 19, 18, 17, 16, 15, 14, 13, 12, 11]

In [24]:
def get_primes(upper_limit):
    if upper_limit == 10:
        return [2, 3, 5, 7]
    if upper_limit == 20:
        return [2, 3, 5, 7, 11, 13, 17, 19]

In [25]:
import numpy as np
def find_number(min_multiple, max_multiple):
    divisors = np.array(get_largest_divisors(min_multiple, max_multiple))
    print(f"Divisors: {divisors}")
    
    primes = get_primes(max_multiple)
    print(f"Primes: {primes}")
    
    prime_division_counter = list()
    for p in primes:
        counter = 0
        nums = divisors
        
        # count the number of times we can divide by any prime in primes with a remainder of 0
        while any(nums % p == 0):
            nums = nums / float(p)
            counter += 1
        prime_division_counter.append(counter)

    prime_division_counter = np.array(prime_division_counter)
    
    # get all the smallest multiple of the number to find by multiplying each prime the counter times we could divide it without remainder
    multiples_for_number = primes ** prime_division_counter
    
    # returns the product of all these multiples
    return np.prod(multiples_for_number)

In [26]:
%%time
find_number(1, 10)

Divisors: [10  9  8  7  6]
Primes: [2, 3, 5, 7]
CPU times: user 1.74 ms, sys: 1.75 ms, total: 3.49 ms
Wall time: 661 µs


2520

In [27]:
%%time
find_number(1, 20)

Divisors: [20 19 18 17 16 15 14 13 12 11]
Primes: [2, 3, 5, 7, 11, 13, 17, 19]
CPU times: user 1.72 ms, sys: 1.77 ms, total: 3.48 ms
Wall time: 973 µs


232792560

# Challenge 6: Sum square difference

The sum of the squares of the first ten natural numbers is, $ 1^2 + 2^2 + ... + 10^2 = 385 $

The square of the sum of the first ten natural numbers is, $ (1 + 2 + ... + 10 )^2 = 55^2 = 3025 $

Hence the difference between the sum of the squares of the first ten natural numbers and the square of the sum is $ 3025 − 385 = 2640 $

Find the difference between the sum of the squares of the first one hundred natural numbers and the square of the sum.

In [28]:
def compute_sum_of_squares(numbers):
    return (numbers ** 2).sum()

In [29]:
%%time
compute_sum_of_squares(np.array(range(1, 11)))

CPU times: user 33 µs, sys: 34 µs, total: 67 µs
Wall time: 69.1 µs


385

In [30]:
def compute_square_of_sum(numbers):
    return numbers.sum() ** 2

In [31]:
%%time
compute_square_of_sum(np.array(range(1, 11)))

CPU times: user 74 µs, sys: 0 ns, total: 74 µs
Wall time: 76.8 µs


3025

In [32]:
%%time
numbers = np.array(range(1, 11))
compute_square_of_sum(numbers) - compute_sum_of_squares(numbers)

CPU times: user 103 µs, sys: 0 ns, total: 103 µs
Wall time: 107 µs


2640

In [33]:
%%time
numbers = np.array(range(1, 101))
compute_square_of_sum(numbers) - compute_sum_of_squares(numbers)

CPU times: user 72 µs, sys: 72 µs, total: 144 µs
Wall time: 148 µs


25164150

# Challenge 7: 10001st prime

By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.

What is the 10 001st prime number?




## Notes

answer: the 10 001st prime is 104_743

We can find the bounds for the nth prime for n >= 6 (https://math.stackexchange.com/questions/1270814/bounds-for-n-th-prime) : $ \log(n) + \log(\log(n)) - 1 < \frac{P_n}{n} < \log(n) + \log(\log(n))$

Afterwards, we generate the prime matrix with Sieve of Eratosthenes algorithm : https://en.wikipedia.org/wiki/Sieve_of_Eratosthenes

In [34]:
import math
def upper_bound(n):
    if n < 6:
        return 11
    else:
        return math.ceil(n * (math.log(n) + math.log(math.log(n))))

In [35]:
%%time
upper_bound(6)

CPU times: user 4 µs, sys: 4 µs, total: 8 µs
Wall time: 9.78 µs


15

In [36]:
%%time
upper_bound(10_001)

CPU times: user 6 µs, sys: 7 µs, total: 13 µs
Wall time: 16.5 µs


114320

In [37]:
def generate_primes(number_limit):
    primes = [True] * (number_limit + 1)
    primes[0] = False
    primes[1] = False
    
    for index, is_prime in enumerate(primes):
        if is_prime:
            yield index
            
            # set all multiples of the prime index to mean not prime
            for n in range(index * index, number_limit + 1, index):
                primes[n] = 0
    return primes

In [38]:
%%time
nth_prime = 6
list(generate_primes(upper_bound(nth_prime)))[nth_prime - 1]

CPU times: user 25 µs, sys: 24 µs, total: 49 µs
Wall time: 52.9 µs


13

In [39]:
%%time
nth_prime = 10_001
list(generate_primes(upper_bound(nth_prime)))[nth_prime - 1]

CPU times: user 29.7 ms, sys: 166 µs, total: 29.8 ms
Wall time: 29.3 ms


104743

# Challenge 8: Largest product in a series

The four adjacent digits in the 1000-digit number that have the greatest product are 9 × 9 × 8 × 9 = 5832.

73167176531330624919225119674426574742355349194934
96983520312774506326239578318016984801869478851843
85861560789112949495459501737958331952853208805511
12540698747158523863050715693290963295227443043557
66896648950445244523161731856403098711121722383113
62229893423380308135336276614282806444486645238749
30358907296290491560440772390713810515859307960866
70172427121883998797908792274921901699720888093776
65727333001053367881220235421809751254540594752243
52584907711670556013604839586446706324415722155397
53697817977846174064955149290862569321978468622482
83972241375657056057490261407972968652414535100474
82166370484403199890008895243450658541227588666881
16427171479924442928230863465674813919123162824586
17866458359124566529476545682848912883142607690042
24219022671055626321111109370544217506941658960408
07198403850962455444362981230987879927244284909188
84580156166097919133875499200524063689912560717606
05886116467109405077541002256983155200055935729725
71636269561882670428252483600823257530420752963450

Find the thirteen adjacent digits in the 1000-digit number that have the greatest product. What is the value of this product?
